source/common/http/filter_manager.cc

```diff
@@ -791,12 +796,36 @@ void FilterManager::decodeMetadata(ActiveStreamDecoderFilter* filter, MetadataMa
       (*entry)->getSavedRequestMetadata()->emplace_back(std::move(metadata_map_ptr));
       return;
     }
-
+    state_.filter_call_state_ |= FilterCallState::DecodeMetadata;
     FilterMetadataStatus status = (*entry)->handle_->decodeMetadata(metadata_map);
-    ASSERT(!hasPreparedLocalReply(), "Sending Local Reply from Metadata is not yet supported.");
+    state_.filter_call_state_ &= ~FilterCallState::DecodeMetadata;
+
     ENVOY_STREAM_LOG(trace, "decode metadata called: filter={} status={}, metadata: {}", *this,
                      (*entry)->filter_context_.config_name, static_cast<uint64_t>(status),
                      metadata_map);
+    if (state_.decoder_filter_chain_aborted_) {
+      // If the decoder filter chain has been aborted, then either:
+      // 1. This filter has sent a local reply from decode metadata.
+      // 2. This filter is the Router filter, and and upstream filter has sent a local reply.
+      ASSERT((status == FilterMetadataStatus::StopIterationForLocalReply) ||
+             (std::next(entry) == decoder_filters_.end()));
```

Comment:
Since we're just checking to see that the filter is the last one, maybe reword the second point to talk about this being a "terminal filter" - the router is one of these but the system allows for any terminal filter to be used at the end of the filter chain

Reply:
Done, thanks for review!